### PR TITLE
fix(56602): JSDoc render with @param Sub-object properties

### LINKED ISSUE #56602: JSDoc render with @param Sub-object properties
### 🔎 Search Terms

"jsdoc", "@param"

### ⏯ Playground Link

_No response_

### 💻 Code

```ts
/**
 * @param {Object} options the args object
 * @param {number} options.alpha first number
 * @param {number} options.bravo second number
 * @param {Object} options.charlie sub-object
 * @param {number} options.charlie.count third number
 * @param {Function} callback the callback function
 * @returns {number}
 */
function addNumbersFromObject(options, callback = null) {
    if (!callback) return options.alpha + options.bravo + options.charlie.count;
    return callback();
}
```

### 🙁 Actual behavior

The render does not display the Sub-object properties of `options.charlie` variable. It's rendering only Object properties of first level.

### 🙂 Expected behavior

The render should display Sub-object properties of `options.charlie` variable like the following image.

image

### Additional information about the issue

Tricky solution:

If you replace the type of `options.charlie` variable in the JSDoc by `any` or `*` , it will display the Sub-object properties in the render.

But if we do that, we will lose the autocompletion feature on `options.charlie` variable.

Recap:

I wish to have at the same time fully JSDoc rendering and autocompletion on an Object variable **AND its Sub-object variables** when it's defined in the JSDoc of my function.

Thanks !

This issue is related to my [previous issue](https://github.com/microsoft/TypeScript/issues/55258).

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #56602

### CODE REVIEW

File: src/services/jsDoc.ts [hunk 1]
```diff
@@ -258,17 +259,21 @@ export function getJsDocTagsFromDeclarations(declarations?: Declaration[], check
         }
         for (const tag of tags) {
             infos.push({ name: tag.tagName.text, text: getCommentDisplayParts(tag, checker) });
-
-            if (isJSDocPropertyLikeTag(tag) && tag.isNameFirst && tag.typeExpression && isJSDocTypeLiteral(tag.typeExpression.type)) {
-                forEach(tag.typeExpression.type.jsDocPropertyTags, propTag => {
-                    infos.push({ name: propTag.tagName.text, text: getCommentDisplayParts(propTag, checker) });
-                });
-            }
+            infos.push(...getJSDocPropertyTagsInfo(tag, checker));
```
Comment (sandersn) [inline src/services/jsDoc.ts:262]:
```suggestion
            infos.push(...getJSDocPropertyTagsInfo(tryGetJSDocPropertyTags(tag), checker)
```

File: src/services/jsDoc.ts [hunk 2]
```diff
@@ -258,17 +259,21 @@ export function getJsDocTagsFromDeclarations(declarations?: Declaration[], check
...
-                });
-            }
+            infos.push(...getJSDocPropertyTagsInfo(tag, checker));
         }
     });
     return infos;
 }
 
+function getJSDocPropertyTagsInfo(node: JSDocTag, checker: TypeChecker | undefined): readonly JSDocTagInfo[] {
+    return flatMap(tryGetJSDocPropertyTags(node), propTag => concatenate([{ name: propTag.tagName.text, text: getCommentDisplayParts(propTag, checker) }], tryGetJSDocPropertyTags(propTag) ? getJSDocPropertyTagsInfo(propTag, checker) : undefined));
```
Comment (sandersn) [inline src/services/jsDoc.ts:269]:
```suggestion
function getJSDocPropertyTagsInfo(nodes: JSDocTag[] | undefined, checker: TypeChecker | undefined): readonly JSDocTagInfo[] {
    return flatMap(nodes, propTag => concatenate([{ name: propTag.tagName.text, text: getCommentDisplayParts(propTag, checker) }],  getJSDocPropertyTagsInfo(tryGetJSDocPropertyTags(propTag), checker) : undefined));

```

**Review by sandersn — CHANGES_REQUESTED** [review]
Without opening this in an editor, I *think* you can make this less redundant by calling tryGetJSDocPropertyTags earlier. But I could be entirely wrong.

File: src/services/jsDoc.ts (hunk 2 above)
Comment (a-tarasyuk) [reply src/services/jsDoc.ts]: Thanks. I've added the requested changed.

**Review by sandersn — APPROVED** [review]

**Pull request merged**